go/store/nbs/store.go

```diff
@@ -1635,26 +1640,154 @@ func (nbs *NomsBlockStore) WriteTableFile(ctx context.Context, fileId string, nu
 }
 
 // AddTableFilesToManifest adds table files to the manifest
-func (nbs *NomsBlockStore) AddTableFilesToManifest(ctx context.Context, fileIdToNumChunks map[string]int) error {
-	var totalChunks int
+func (nbs *NomsBlockStore) AddTableFilesToManifest(ctx context.Context, fileIdToNumChunks map[string]int, getAddrs chunks.GetAddrsCurry) error {
+	return nbs.addTableFilesToManifest(ctx, fileIdToNumChunks, getAddrs, nbs.refCheck)
+}
+
+// A small helper which returns a composite refCheck function combining |base| and a |hasMany| call against each source in |css|.
+func getRefCheck(nbs *NomsBlockStore, base func([]hasRecord) (hash.HashSet, error), css chunkSourceSet) func([]hasRecord) (hash.HashSet, error) {
+	if base == nil {
+		return nil
+	}
+	return func(recs []hasRecord) (hash.HashSet, error) {
+		nbs.mu.Lock()
+		absent, err := base(recs)
+		nbs.mu.Unlock()
+		if err != nil {
+			return nil, err
+		}
+		if len(absent) == 0 {
+			return nil, nil
+		}
+
+		// Still some missing. Check the chunkSourceSet.
+		for _, cs := range css {
+			remaining, _, err := cs.hasMany(recs, nil)
+			if err != nil {
+				return nil, err
+			}
+			if !remaining {
+				return nil, nil
+			}
+		}
+
+		ret := hash.HashSet{}
+		for _, r := range recs {
+			if !r.has {
+				ret.Insert(*r.a)
+			}
+		}
+		return ret, nil
+	}
+}
+
+// For each chunk in sources, walk all its addresses with |getAddrs| and ensure that none are missing from a call to |refCheck|.
```

Comment:
so basically, the chunkstores are self-consistent and do not include any intermediate nodes with missing refs?